File: app/models/business.py
from datetime import UTC, date, datetime, time

from sqlalchemy import (
    JSON,
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    Time,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


def utc_now() -> datetime:
    return datetime.now(UTC)


class BusinessProfile(Base):
    __tablename__ = "business_profiles"
    __table_args__ = (UniqueConstraint("company_id", name="uq_business_profiles_company"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="CASCADE"), index=True)
    business_name: Mapped[str] = mapped_column(String(255))
    phone: Mapped[str | None] = mapped_column(String(30), nullable=True)
    address: Mapped[str | None] = mapped_column(String(255), nullable=True)
    city: Mapped[str | None] = mapped_column(String(120), nullable=True)
    neighborhood: Mapped[str | None] = mapped_column(String(120), nullable=True)
    delivery_fee: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    estimated_delivery_time: Mapped[str | None] = mapped_column(String(120), nullable=True)
    accepts_pickup: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    payment_methods: Mapped[list[str] | None] = mapped_column(JSON, nullable=True)
    welcome_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    out_of_hours_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        server_default=func.now(),
    )

    company = relationship("Company", back_populates="business_profile")


class ProductCategory(Base):
    __tablename__ = "product_categories"
    __table_args__ = (Index("ix_product_categories_company_active", "company_id", "active"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(255), index=True)
    active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        server_default=func.now(),
    )

    company = relationship("Company", back_populates="product_categories")
    products = relationship("Product", back_populates="category")


class Product(Base):
    __tablename__ = "products"
    __table_args__ = (
        Index("ix_products_company_active_featured", "company_id", "active", "featured"),
        Index("ix_products_company_category_order", "company_id", "category_id", "display_order"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="CASCADE"), index=True)
    category_id: Mapped[int | None] = mapped_column(
        ForeignKey("product_categories.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(255), index=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    price: Mapped[float] = mapped_column(Numeric(10, 2))
    promotional_price: Mapped[float | None] = mapped_column(Numeric(10, 2), nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    featured: Mapped[bool] = mapped_column(Boolean, default=False, index=True)
    display_order: Mapped[int] = mapped_column(Integer, default=0, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        server_default=func.now(),
    )

    company = relationship("Company", back_populates="products")
    category = relationship("ProductCategory", back_populates="products")
    addons = relationship("ProductAddon", back_populates="product", cascade="all, delete-orphan")


class ProductAddon(Base):
    __tablename__ = "product_addons"
    __table_args__ = (Index("ix_product_addons_company_product_active", "company_id", "product_id", "active"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="CASCADE"), index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(255))
    price: Mapped[float] = mapped_column(Numeric(10, 2))
    active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        server_default=func.now(),
    )

    company = relationship("Company", back_populates="product_addons")
    product = relationship("Product", back_populates="addons")


class BusinessHour(Base):
    __tablename__ = "business_hours"
    __table_args__ = (
        UniqueConstraint("company_id", "day_of_week", name="uq_business_hours_company_day"),
        Index("ix_business_hours_company_active", "company_id", "active"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="CASCADE"), index=True)
    day_of_week: Mapped[int] = mapped_column(Integer, index=True)
    open_time: Mapped[time] = mapped_column(Time())
    close_time: Mapped[time] = mapped_column(Time())
    active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        server_default=func.now(),
    )

    company = relationship("Company", back_populates="business_hours_entries")


class Promotion(Base):
    __tablename__ = "promotions"
    __table_args__ = (Index("ix_promotions_company_active_dates", "company_id", "active", "start_date", "end_date"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)
    start_date: Mapped[date | None] = mapped_column(Date(), nullable=True)
    end_date: Mapped[date | None] = mapped_column(Date(), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        server_default=func.now(),
    )

    company = relationship("Company", back_populates="promotions")

File: frontend/src/lib/validations/business.ts
import { z } from "zod";

export const businessProfileSchema = z.object({
  business_name: z.string().min(2, "Informe o nome do negócio."),
  phone: z.string().optional().or(z.literal("")),
  address: z.string().optional().or(z.literal("")),
  city: z.string().optional().or(z.literal("")),
  neighborhood: z.string().optional().or(z.literal("")),
  delivery_fee: z.coerce.number().min(0).nullable().optional(),
  estimated_delivery_time: z.string().optional().or(z.literal("")),
  accepts_pickup: z.boolean(),
  payment_methods_text: z.string().optional().or(z.literal("")),
  welcome_message: z.string().optional().or(z.literal("")),
  out_of_hours_message: z.string().optional().or(z.literal(""))
});

export type BusinessProfileSchema = z.infer<typeof businessProfileSchema>;

export const categorySchema = z.object({
  name: z.string().min(2, "Informe o nome da categoria."),
  active: z.boolean()
});

export type CategorySchema = z.infer<typeof categorySchema>;

export const productSchema = z.object({
  category_id: z.coerce.number().nullable().optional(),
  name: z.string().min(2, "Informe o nome do produto."),
  description: z.string().optional().or(z.literal("")),
  price: z.coerce.number().min(0, "Informe um preço válido."),
  promotional_price: z.coerce.number().min(0).nullable().optional(),
  active: z.boolean(),
  featured: z.boolean(),
  display_order: z.coerce.number().min(0)
});

export type ProductSchema = z.infer<typeof productSchema>;

export const addonSchema = z.object({
  product_id: z.coerce.number().min(1, "Selecione um produto."),
  name: z.string().min(2, "Informe o nome do adicional."),
  price: z.coerce.number().min(0, "Informe um preço válido."),
  active: z.boolean()
});

export type AddonSchema = z.infer<typeof addonSchema>;

export const businessHourSchema = z.object({
  day_of_week: z.coerce.number().min(0).max(6),
  open_time: z.string().min(1, "Informe a hora de abertura."),
  close_time: z.string().min(1, "Informe a hora de fechamento."),
  active: z.boolean()
});

export type BusinessHourSchema = z.infer<typeof businessHourSchema>;

export const promotionSchema = z.object({
  title: z.string().min(2, "Informe o título da promoção."),
  description: z.string().optional().or(z.literal("")),
  active: z.boolean(),
  start_date: z.string().optional().or(z.literal("")),
  end_date: z.string().optional().or(z.literal(""))
});

export type PromotionSchema = z.infer<typeof promotionSchema>;
